File: app/src/main/java/com/nikola/jakshic/truesight/viewModel/SearchViewModel.java
package com.nikola.jakshic.truesight.viewModel;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.nikola.jakshic.truesight.data.local.SearchHistoryDao;
import com.nikola.jakshic.truesight.model.Player;
import com.nikola.jakshic.truesight.model.SearchHistory;
import com.nikola.jakshic.truesight.repository.PlayerRepository;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

public class SearchViewModel extends ViewModel {

    private MutableLiveData<List<Player>> playerList;
    private MutableLiveData<List<SearchHistory>> historyList;
    private PlayerRepository repository;
    private SearchHistoryDao searchHistoryDao;
    private MutableLiveData<Boolean> loading;

    @Inject
    public SearchViewModel(PlayerRepository repository, SearchHistoryDao searchHistoryDao) {
        this.repository = repository;
        this.searchHistoryDao = searchHistoryDao;
        playerList = new MutableLiveData<>();
        historyList = new MutableLiveData<>();
        loading = new MutableLiveData<>();
        loading.setValue(false);
    }

    public void getAllQueries() {
        historyList.setValue(searchHistoryDao.getAllQueries());
    }

    public MutableLiveData<List<SearchHistory>> getSearchHistory() {
        return historyList;
    }

    public void getQueries(String query) {
        historyList.setValue(searchHistoryDao.getQuery(query));
    }

    public void saveQuery(SearchHistory searchHistory) {
        searchHistoryDao.insertQuery(searchHistory);
    }

    public MutableLiveData<List<Player>> getPlayers() {
        return playerList;
    }

    public void fetchPlayers(String name) {
        repository.fetchPlayers(playerList, loading, name);
    }

    public MutableLiveData<Boolean> isLoading() {
        return loading;
    }

    public void clearList() {
        playerList.setValue(new ArrayList<>());
    }
}

File: app/src/main/java/com/nikola/jakshic/dagger/ui/search/SearchViewModel.kt
package com.nikola.jakshic.dagger.ui.search

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.nikola.jakshic.dagger.data.local.SearchHistoryDao
import com.nikola.jakshic.dagger.repository.PlayerRepository
import com.nikola.jakshic.dagger.ui.ScopedViewModel
import com.nikola.jakshic.dagger.ui.Status
import com.nikola.jakshic.dagger.vo.Player
import com.nikola.jakshic.dagger.vo.SearchHistory
import kotlinx.coroutines.*
import javax.inject.Inject

class SearchViewModel @Inject constructor(
        private val dao: SearchHistoryDao,
        private val repository: PlayerRepository
) : ScopedViewModel() {

    private var job = Job()
    private val scope = CoroutineScope(Dispatchers.Main + job)

    private val _playerList = MutableLiveData<List<Player>>()
    val playerList: LiveData<List<Player>>
        get() = _playerList

    private val _historyList = MutableLiveData<List<SearchHistory>>()
    val historyList: LiveData<List<SearchHistory>>
        get() = _historyList

    private val _status = MutableLiveData<Status>()
    val status: LiveData<Status>
        get() = _status

    private val onSuccess: (List<Player>) -> Unit = {
        _status.value = Status.SUCCESS
        _playerList.value = it
    }
    private val onError: () -> Unit = { _status.value = Status.ERROR }

    fun getAllQueries() {
        launch {
            val list = withContext(Dispatchers.IO) { dao.getAllQueries() }
            _historyList.value = list
        }
    }

    fun getQueries(query: String) {
        launch {
            val list = withContext(Dispatchers.IO) { dao.getQuery(query) }
            _historyList.value = list
        }
    }

    fun saveQuery(item: SearchHistory) {
        launch {
            withContext(Dispatchers.IO) { dao.insertQuery(item) }
        }
    }

    fun fetchPlayers(name: String) {
        // Users can hit the search button multiple times
        // So we need to cancel previous call
        job.cancelChildren()

        scope.launch {
            _status.value = Status.LOADING
            repository.fetchPlayers(name, onSuccess, onError)
        }
    }

    fun clearList() {
        _playerList.value = ArrayList()
    }

    override fun onCleared() {
        super.onCleared()
        job.cancel()
    }
}